lib/outdated.js

```diff
@@ -1,437 +1,244 @@
+    const peer = pkg.peerDependencies
+    const optional = pkg.optionalDependencies
+
+    const type = (
+      (prod && prod[name] && 'dependencies') ||
+      (dev && dev[name] && 'devDependencies') ||
+      (peer && peer[name] && 'peerDependencies') ||
+      (optional && optional[name] && 'optionalDependencies') ||
+      'dependencies'
     )
-  ) {
-    Object.keys(tree.missingDevDeps).forEach(function (name) {
-      deps.push({
-        package: { name: name },
-        path: tree.path,
-        parent: tree,
-        isMissing: true
+
+    // deps different from prod not currently
+    // on disk are not included in the output
+    if (!current && type !== 'dependencies') return
+
+    const expected = pkg[type] && pkg[type][name]
+
+    try {
+      const packument = await pacote.packument(spec, {
+        fullMetadata: true,
+        preferOnline: true
       })
-      if (!types[name]) {
-        types[name] = 'devDependencies'
+      const wanted = pickManifest(packument, expected)
+      const latest = pickManifest(packument, 'latest')
+
+      if (
+        !current ||
+        current !== wanted.version ||
+        wanted.version !== latest.version
+      ) {
+        list.push({
+          name,
+          path,
+          type,
+          current,
+          homepage: packument.homepage,
+          wanted: wanted.version,
+          latest: latest.version,
+          location
+        })
       }
-    })
-  }
-
-  if (opts['save-dev']) {
-    deps = deps.filter(function (dep) { return pkg.devDependencies[moduleName(dep)] })
-    deps.forEach(function (dep) {
-      types[moduleName(dep)] = 'devDependencies'
-    })
-  } else if (opts.save) {
-    // remove optional dependencies from dependencies during --save.
-    deps = deps.filter(function (dep) { return !pkg.optionalDependencies[moduleName(dep)] })
-  } else if (opts['save-optional']) {
-    deps = deps.filter(function (dep) { return pkg.optionalDependencies[moduleName(dep)] })
-    deps.forEach(function (dep) {
-      types[moduleName(dep)] = 'optionalDependencies'
-    })
-  }
-  var doUpdate = dev || (
-    !prod &&
-    !Object.keys(parentHas).length &&
-    !opts.global
-  )
-  if (doUpdate) {
-    Object.keys(pkg.devDependencies || {}).forEach(function (k) {
-      if (!(k in parentHas)) {
-        deps[k] = pkg.devDependencies[k]
-        types[k] = 'devDependencies'
+    } catch (err) {
+      // silently catch and ignore ETARGET, E403 &
+      // E404 errors, deps are just skipped
+      if (!(
+        err.code === 'ETARGET' ||
+        err.code === 'E403' ||
+        err.code === 'E404')
+      ) {
+        throw (err)
       }
-    })
+    }
+
   }
 
-  var has = Object.create(parentHas)
-  tree.children.forEach(function (child) {
-    if (moduleName(child) && child.package.private) {
-      deps = deps.filter(function (dep) { return dep !== child })
+  if (deps.length !== 0) {
+    // specific deps
+    for (let i = 0; i < deps.length; i++) {
```

Comment:
We can go faster here by parallelizing rather than waiting for each item to finish before fetching the next packument.

```js
const waiting = []
if (...) {
  for (all the nodes) {
    waiting.push(getOutdatedInfo(name, node))
  }
} else { ... }

await Promise.all(waiting)
```

Reply:
If we do that though, we'll also want to sort the list after the `Promise.all()`, since they might resolve in any random order.